### PR TITLE
Reference File Upload

### PR DESCRIPTION
**Pull request by BlobbyBlobbfish**
Uploading of Reference File for review from Team 5.

### CODE REVIEW

File: scripting/FileParsing.js [hunk 1]
```diff
@@ -1 +1,141 @@
...
+          if(fileType === "csv"){
+              //if its cvs split it into an array based upon new lines, each element of array is a row in file
+              let valuesArray = stringResults.split(/\r?\n|\r/);
+              //passes it to handeling for cvs files
+              return csvHandeling(valuesArray);
+          }
+          else if(fileType === "json"){
+              //if its a json file then use built in parsing and return parsed object
+              let jsonObj = JSON.parse(stringResults);
+              return jsonObj;
```
Comment (Norwegia) [inline scripting/FileParsing.js:36]: why make a variable and not just return the JSON.parse() result?

File: scripting/FileParsing.js [hunk 2]
```diff
@@ -1 +1,141 @@
...
+
+      //initiates file parsing
+      fileParse(stringResults, fileType);
+      });
+      fileReader.readAsText(file); //reads content as text to activate load for event listener
+  }
+  function fileParse(stringResults, fileType){
+          if(fileType === "csv"){
+              //if its cvs split it into an array based upon new lines, each element of array is a row in file
+              let valuesArray = stringResults.split(/\r?\n|\r/);
```
Comment (Norwegia) [inline scripting/FileParsing.js:29]: Aren't CSVs seperated by commas and not new lines so shouldn't you be spliting by that instead?

File: scripting/FileParsing.js [hunk 3]
```diff
@@ -1 +1,141 @@
...
+          else{
+              //yeild error because wrong file type
+              alert("incorrect file type: please enter a JSON or CVS file and try again.");
+          }
+  }
+  //method to replace characters in a string based upon index
+  String.prototype.replaceAt = function(index, replacement) {
+  	if (index >= this.length) {
+  		return this.valueOf();
+  	}
```
Comment (Norwegia) [inline scripting/FileParsing.js:46]: can be simplified by just writing: if (index >= this.length) return this.valueOf();

File: scripting/FileParsing.js [hunk 4]
```diff
@@ -1 +1,141 @@
...
+                  isComma = true;
+              }
+              else if(array[y].charAt(x) == "\""  && isComma){
+                  isComma = false;
+              }
+              if(isComma && array[y].charAt(x) == ","){
+                  array[y] = array[y].replaceAt(x,"^");
+              }
+          }
+      }
```
Comment (Norwegia) [inline scripting/FileParsing.js:67]: Why is this necessary? Can't you just seperate by commas in the orginal array being passed in and clear whitespace if it is a problem?

File: scripting/FileParsing.js [hunk 5]
```diff
@@ -1 +1,141 @@
...
+      //breaks array into multidimentional array - essentially a grid same as the excel doc
+      for(let x = 0; x < array.length; x++){
+          array[x] = array[x].split(",");
+      }
+      //finds which column has mod_id and deletes all the garbage before it, deletes everything between mod_id and AM, deletes everything after the last loci
```
Comment (Norwegia) [inline scripting/FileParsing.js:72]: Seems this could be split into a different function

File: scripting/FileParsing.js [hunk 6]
```diff
@@ -1 +1,141 @@
...
+      for(x = 0; x < array[0].length; x++){
+          if(array[0][x].slice(0,3) == "mod" && array[0][x] != "mod_id"){
+              correctColumn = x;
+          }
+      }
+      if(correctColumn > 0){
+          for(let y = 0; y < array.length; y++){
+              array[y].splice(correctColumn,array[0].length - 1);
+          }
+      }
```
Comment (Norwegia) [inline scripting/FileParsing.js:98]: I am a little bit confused by this code - might want to add some more comments. It also seems like you are iterating over the file quite bit, which seems a little unnecessary.

File: scripting/FileParsing.js [hunk 7]
```diff
@@ -1 +1,141 @@
...
+  //method to replace characters in a string based upon index
+  String.prototype.replaceAt = function(index, replacement) {
+  	if (index >= this.length) {
+  		return this.valueOf();
+  	}
+
+  	return this.substring(0, index) + replacement + this.substring(index + 1);
+  }
+
+  function csvHandeling(array){
```
Comment (Norwegia) [inline scripting/FileParsing.js:51]: array seems like maybe a little too generic of a variable name, maybe should be renamed to something that includes what the array is supposed to store

File: scripting/FileParsing.js (hunk 1 above)
Comment (BlobbyBlobbfish) [reply scripting/FileParsing.js:36]: Good point

File: scripting/FileParsing.js [hunk 8]
```diff
@@ -1 +1,141 @@
...
+                      array[y][x] = array[y][x].substr(0, array[y][x].length - 1);
+                  }
+              }
+          }
+      }
+      for(y = 0; y < array.length; y++){
+          for(x = 1; x < array[y].length; x++){
+              array[y][x] = array[y][x].split("^");
+          }
+      }
```
Comment (Norwegia) [inline scripting/FileParsing.js:114]: Again, I haven't written this code so I don't understand all of how it is meant to work, but you seem to be iterating a whole lot more than what should be necessary.

File: scripting/FileParsing.js [hunk 9]
```diff
@@ -1 +1,141 @@
...
+      //object maker
+      let objArray = [];
```
Comment (Norwegia) [inline scripting/FileParsing.js:116]: Again, this function seems very long, might want to split this object maker into another function

File: scripting/FileParsing.js [hunk 10]
```diff
@@ -1 +1,141 @@
...
+          loci  = {};
+          for(x = 1; x < array[y].length; x++){
+              loci[array[0][x]] = [array[y][x]];
+          }
+          tempObj.loci = loci;
+          tempObj.modelIdentification = array[y][0];
+          objArray[y - 1] = tempObj;
+      }
+      for(let y = 0; y < objArray.length; y++){
+          if(objArray[y].modelIdentification === "" || objArray[y].modelIdentification === undefined || objArray[y].modelIdentification === null || objArray[y].modelIdentification === "<empty string>"){
```
Comment (Norwegia) [inline scripting/FileParsing.js:129]: Very long logic statement, maybe should be written in multiple lines or make a function for it.

File: scripting/FileParsing.js [hunk 11]
```diff
@@ -1 +1,140 @@
...
+              objArray.splice(y,1);
+          }
+      }
+      for(let y = 0; y < objArray.length; y++){
+          if(objArray[y].modelIdentification === "" || objArray[y].modelIdentification === undefined || objArray[y].modelIdentification === null || objArray[y].modelIdentification === "<empty string>"){
```
Comment (Norwegia) [inline scripting/FileParsing.js:134]: Again, long logic statement

Comment (Norwegia) [reply scripting/FileParsing.js:134]: Also, it looks like it seems you are doing the  same thing twice
